x-pack/plugins/licensing/server/services/feature_usage_service.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/** @public */
+export interface FeatureUsageServiceSetup {
+  /**
+   * Register a feature to be able to notify of it's usages using the {@link FeatureUsageServiceStart | service start contract}.
+   */
+  register(featureName: string): void;
+}
+
+/** @public */
+export interface FeatureUsageServiceStart {
+  /**
+   * Notify of a registered feature usage at given time.
+   * If `usedAt` is not specified, it will use the current time instead.
+   */
+  notifyUsage(featureName: string, usedAt?: number): void;
+  /**
+   * Return a map containing last usage timestamp for all features.
+   * Features that were not used yet do not appear in the map.
+   */
+  getLastUsages(): ReadonlyMap<string, number>;
+  /**
+   * Clear all usage records from the service.
+   */
+  clear(): void;
+}
+
+export class FeatureUsageService {
+  private readonly features: string[] = [];
+  private readonly lastUsages = new Map<string, number>();
+
+  public setup(): FeatureUsageServiceSetup {
+    return {
+      register: featureName => {
+        if (this.features.includes(featureName)) {
+          throw new Error(`Feature '${featureName}' has already been registered.`);
+        }
+        this.features.push(featureName);
+      },
+    };
+  }
+
+  public start(): FeatureUsageServiceStart {
+    return {
+      notifyUsage: (featureName, usedAt = Date.now()) => {
+        if (!this.features.includes(featureName)) {
+          throw new Error(`Feature '${featureName}' is not registered.`);
+        }
+        const currentValue = this.lastUsages.get(featureName) ?? 0;
+        this.lastUsages.set(featureName, Math.max(usedAt, currentValue));
+      },
+      getLastUsages: () => new Map(this.lastUsages.entries()),
```

Comment:
Should this return `null` values for features that were registered but have not been used? As it is right now, features that have not been used do not show up in the API response at all. I'm not sure what consumers of this expect, but it may be a bit clearer if all the tracked features show in this list, regardless of usage.

Reply:
I'm following @rjernst 'specifications' from https://github.com/elastic/kibana/issues/60984#issuecomment-608563910 here. non-used features does not appear in this API apparently.

I'm unsure what liberty we have on that, as the goal seems to be mirroring ES' WIP api?

Reply:
why we need `features` list at all in this case? Can we remove`register` call?

Reply:
> why we need features list in this case at all? Can we remove register call?

That was my first thought as well. If we don't need to return the list of all features (even if unused) then we could probably remove it. Though it does make it a bit safer that plugins can't just shove random strings into the `notifyUsage` API.

Maybe we could get away with a union string type for the allowed strings, though that would mean plugins have to update these types in order to add new usage tracking.

Reply:
> why we need features list in this case at all? Can we remove register call?

There is no functional reasons for this call atm. I just added it because it felt more KP compliant for plugins to explicitly register every licensed feature they wanted to notify usages of. If we think this is useless, I can definitely remove that part.

> Maybe we could get away with a union string type for the allowed strings

Mixed on that. In addition to `that would mean plugins have to update these types in order to add new usage tracking`, this also kinda breaks the feature for 3rd party plugins, as they wouldn't be able to register new features, even if realistically it's probably a non-issue.

Reply:
> Though it does make it a bit safer that plugins can't just shove random strings into the notifyUsage API.

What prevent them from passing a random string to `register` as well?

Reply:
Nothing, but at least they can only notify usage of the arbitrary features that were registered.